### PR TITLE
[Example] Add apply_adam_w and masked_scale examples

### PR DESCRIPTION
**Pull request by qxyaba0**
Add two new operator examples under `examples/cann-bench`:

- `apply_adam_w/example_apply_adam_w.py`: Decoupled AdamW optimizer update kernel on Ascend NPU.
- `masked_scale/example_masked_scale.py`: Elementwise masked-scale kernel with multiple dtype-specific code paths.

新增 apply_adam_w 和 masked_scale 算子，以下为两个算子在在A3上运行CANN-BENCH测试(对比CANN-BENCH内置的baseline)以及在线测评的加速比：

| 算子 | 自测加速比 | 在线测评加速比 |
|------|--------------|-------------------|
| apply_adam_w | 1.12x | 1.086× |
| masked_scale | 1.28x | 1.232× |

具体算子测评报告如下：

# 算子评测报告

**评测代号**: cann_final_eval_20260821_152602
**评测时间**: 2026-08-21T15:26:02.582196
**设备**: npu:0
**框架版本**: V1.0.0
**评测集版本**: tasks-v1.0.0

## 概览

| 指标 | 数值 |
|------|------|
| 评测算子数 | 1 |
| 总用例数 | 20 |
| 通过用例数 | 20 |
| 失败用例数 | 0 |
| 通过率 | 100.00% |
| 综合得分 | 76.29 |

### ApplyAdamW（level2/apply_adam_w）

| 指标 | 数值 |
|------|------|
| 用例数 | 20 |
| 通过数 | 20 |
| 失败数 | 0 |
| 通过率 | 100.00% |
| 平均加速比 | 1.12x |
| 编译/运行得分 | 20.00 |
| 精度得分 | 30.00 |
| 性能得分 | 26.29 |
| 得分 | 76.29 |

| 用例ID | 状态 | 耗时(μs) | 加速比 | 精度误差 |
|--------|------|----------|--------|----------|
| level2/apply_adam_w_1 | ✅ | 18.56 | 2.05x | 0.000000 |
| level2/apply_adam_w_2 | ✅ | 42.48 | 1.21x | 0.000244 |
| level2/apply_adam_w_3 | ✅ | 141.78 | 1.73x | 0.000000 |
| level2/apply_adam_w_4 | ✅ | 266.00 | 1.21x | 0.000001 |
| level2/apply_adam_w_5 | ✅ | 287.73 | 1.07x | 0.003906 |
| level2/apply_adam_w_6 | ✅ | 25.40 | 1.36x | 0.000000 |
| level2/apply_adam_w_7 | ✅ | 31.74 | 1.21x | 0.000000 |
| level2/apply_adam_w_8 | ✅ | 27.42 | 1.08x | 0.000000 |
| level2/apply_adam_w_9 | ✅ | 1106.14 | 0.69x | 0.062500 |
| level2/apply_adam_w_10 | ✅ | 35.16 | 1.03x | 0.003906 |
| level2/apply_adam_w_11 | ✅ | 27.76 | 1.03x | 0.000000 |
| level2/apply_adam_w_12 | ✅ | 26.06 | 1.31x | 0.000000 |
| level2/apply_adam_w_13 | ✅ | 258.83 | 0.65x | 0.000000 |
| level2/apply_adam_w_14 | ✅ | 61.90 | 0.73x | 0.000000 |
| level2/apply_adam_w_15 | ✅ | 33.22 | 1.16x | 0.000000 |
| level2/apply_adam_w_16 | ✅ | 41.14 | 1.17x | 0.000002 |
| level2/apply_adam_w_17 | ✅ | 52.70 | 0.74x | 0.250000 |
| level2/apply_adam_w_18 | ✅ | 53.88 | 0.96x | 0.000000 |
| level2/apply_adam_w_19 | ✅ | 40.48 | 1.19x | 0.000000 |
| level2/apply_adam_w_20 | ✅ | 173.48 | 0.91x | 0.000004 |

## 算子流水图
image

# 算子评测报告

**评测代号**: maskedscale_eval_20260821_171608
**评测时间**: 2026-08-21T17:18:33.800847
**设备**: npu:0
**框架版本**: V1.0.0
**评测集版本**: tasks-v1.0.0

## 概览

| 指标 | 数值 |
|------|------|
| 评测算子数 | 1 |
| 总用例数 | 20 |
| 通过用例数 | 20 |
| 失败用例数 | 0 |
| 通过率 | 100.00% |
| 综合得分 | 81.21 |

### MaskedScale（level1/masked_scale）

| 指标 | 数值 |
|------|------|
| 用例数 | 20 |
| 通过数 | 20 |
| 失败数 | 0 |
| 通过率 | 100.00% |
| 平均加速比 | 1.28x |
| 编译/运行得分 | 20.00 |
| 精度得分 | 30.00 |
| 性能得分 | 31.21 |
| 得分 | 81.21 |

| 用例ID | 状态 | 耗时(μs) | 加速比 | 精度误差 |
|--------|------|----------|--------|----------|
| level1/masked_scale_1 | ✅ | 7.28 | 2.13x | 0.000000 |
| level1/masked_scale_2 | ✅ | 25.66 | 1.47x | 0.000000 |
| level1/masked_scale_3 | ✅ | 101.36 | 0.87x | 0.000000 |
| level1/masked_scale_4 | ✅ | 550.11 | 1.52x | 0.000000 |
| level1/masked_scale_5 | ✅ | 519.43 | 2.57x | 0.000000 |
| level1/masked_scale_6 | ✅ | 17.58 | 0.83x | 0.000000 |
| level1/masked_scale_7 | ✅ | 15.76 | 1.02x | 0.000000 |
| level1/masked_scale_8 | ✅ | 26.74 | 0.82x | 0.000008 |
| level1/masked_scale_9 | ✅ | 801.30 | 0.72x | 0.500000 |
| level1/masked_scale_10 | ✅ | 16.54 | 0.91x | 0.000000 |
| level1/masked_scale_11 | ✅ | 22.34 | 0.86x | 0.000000 |
| level1/masked_scale_12 | ✅ | 9.28 | 1.20x | 0.000000 |
| level1/masked_scale_13 | ✅ | 16.94 | 5.18x | 0.000000 |
| level1/masked_scale_14 | ✅ | 20.02 | 1.21x | 0.000000 |
| level1/masked_scale_15 | ✅ | 20.58 | 0.89x | 0.000000 |
| level1/masked_scale_16 | ✅ | 32.86 | 0.54x | 0.000000 |
| level1/masked_scale_17 | ✅ | 17.16 | 1.09x | 0.000000 |
| level1/masked_scale_18 | ✅ | 41.56 | 0.64x | 0.000000 |
| level1/masked_scale_19 | ✅ | 39.80 | 0.63x | 0.062500 |
| level1/masked_scale_20 | ✅ | 123.52 | 0.56x | 0.000000 |

## 算子流水图
image

### CODE REVIEW

**Review by fengz72 — DISMISSED** [review]

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 1]
```diff
@@ -0,0 +1,1831 @@
...
+            epsilon,
+            tl_dtype,
+        )
+        y_2d = kernel(var_2d, grad_2d, m_2d, v_2d)
+
+    return y_2d.reshape(original_shape)
+
+
+# ---------------------------------------------------------------------------
+# Golden reference (matches cann-bench golden.py exactly)
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py]: 此PR中无此文件

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 2]
```diff
@@ -0,0 +1,1831 @@
...
+    v,
+    lr,
+    beta1,
+    beta2,
+    weight_decay,
+    epsilon=1e-8,
+    step=1,
+    maximize=False,
+):
+    """PyTorch reference implementation (consistent with cann-bench golden.py).
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py]: 此PR中无此文件

File: examples/cann-bench/masked_scale/example_masked_scale.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
+import argparse
+
+import tilelang
+import tilelang.language as T
+import torch
+
+tilelang.cache.clear_cache()
```
Comment (fengz72) [inline examples/cann-bench/masked_scale/example_masked_scale.py]: tilelang.cache.clear_cache() 是删除整个磁盘 kernel cache, 破坏性太大. cache 报错的bug已在 https://github.com/tile-ai/tilelang-ascend/commit/f55c24f7a3e471a582e9a4c8982e3586fcd5043a 修复, 建议删除此代码

File: examples/cann-bench/masked_scale/example_masked_scale.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+                    if scale != 1.0:
+                        T.tile.mul(x_load, x_load, scale)
+
+                    T.copy(x_load, Y[row_start, col_start])
+
+    return main
+
+
+parser = argparse.ArgumentParser(description="NPU Kernel Compilation")
```
Comment (fengz72) [inline examples/cann-bench/masked_scale/example_masked_scale.py]: argparse、kernel 编译、测试循环建议移进 `__main__`

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 5]
```diff
@@ -0,0 +1,1831 @@
...
+            eff_beta1,
+            eff_ombeta1,
+            eff_beta2,
+            eff_ombeta2,
+            lr_signed,
+            weight_decay,
+            epsilon,
+            tl_dtype,
+        )
+        if key not in _pipeline_lowprec_cache:
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py]: 使用tilelang框架的cache, 不要使用模块级cache

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 6]
```diff
@@ -0,0 +1,1831 @@
...
+CAST_MODE_LOW2HIGH = "CAST_NONE"
+CAST_MODE_HIGH2LOW = "CAST_RINT"
+
+_CORE_NUM = 24
+_UB_LIMIT_BYTES = 160 * 1024
+_NUM_INPUTS = 4  # var, grad, m, v
+_SCALAR_PAD = 8  # 7 scalars + 1 padding (eff_beta1, eff_ombeta1, eff_beta2,
+#   eff_ombeta2, lr_signed, weight_decay, epsilon, 0.0)
+
+_pipeline_fp32_cache = {}
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py]: 使用tilelang框架的cache, 不要使用模块级cache

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 7]
```diff
@@ -0,0 +1,1831 @@
...
+_pipeline_lowprec_cache = {}
+_barrier_kernel_cache = {}
+_scalar_tensor_cache = {}
+
+
+def torch_dtype_to_tl(dtype):
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py:76]: 可以改为map, 更简洁直观

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 8]
```diff
@@ -0,0 +1,1831 @@
...
+        return "float16"
+    elif dtype == torch.bfloat16:
+        return "bfloat16"
+    elif dtype == torch.float32:
+        return "float"
+    else:
+        raise ValueError(f"Unsupported dtype: {dtype}")
+
+
+def _compute_2d_shape(shape):
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py:82]: 可以简化
```python
def _compute_2d_shape(shape):
    """Reshape to 2D (M, N): merge trailing dims until N >= 256."""
    total = math.prod(shape)
    N = 1
    for s in reversed(shape):
        N *= s
        if N >= 256:
            break
    return total // N, N
```

File: examples/cann-bench/apply_adam_w/example_apply_adam_w.py [hunk 9]
```diff
@@ -0,0 +1,1831 @@
...
+    return (eff_beta1, eff_ombeta1, eff_beta2, eff_ombeta2, lr_signed)
+
+
+# ---------------------------------------------------------------------------
+# Pipeline kernel: fp32 path (no cal buffers, compute directly on input UB)
+# ---------------------------------------------------------------------------
+
+
+@tilelang.jit(out_idx=[4], pass_configs=_PIPELINE_CONFIGS)
+def _adam_w_pipeline_fp32(
```
Comment (fengz72) [inline examples/cann-bench/apply_adam_w/example_apply_adam_w.py:292]: 四个kernel的计算流程有部分一致, 可以使用`T.micro`提取为函数

**Comment by MiaoHaifeng** [comment]
approve

**Comment by MiaoHaifeng** [comment]
lgtm

**Review by MiaoHaifeng — APPROVED** [review]
